### PR TITLE
templatazing SubstitutionVisitor

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h [hunk 1]
```diff
@@ -31,28 +31,61 @@ namespace Antares::Solver::Visitors
...
+                                   [&node](auto* x) { return *x == *node; });
+            if (it != ctx_.variables.end())
+            {
+                return *it;
+            }
+
+            else
+            {
+                return CloneVisitor::visit(node);
+            }
```
Comment (guilpier-code) [inline src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h]:
```C++
Nodes::Node* visit(const NodeToSubstitute* node) override
{
    ...
    if (it != ctx_.variables.end())
    {
        return *it;
    }
    return CloneVisitor::visit(node);
}
```

File: src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp [hunk 2]
```diff
@@ -36,29 +37,37 @@ using namespace Antares::Solver::Visitors;
 
 BOOST_AUTO_TEST_SUITE(_SubstitutionVisitor_)
 
-std::pair<Nodes::ComponentVariableNode*, Nodes::ComponentVariableNode*> fillContext(
-  SubstitutionContext& ctx,
+template<typename NodeToSubstitute>
+concept IsComponentVariableorPortField = std::same_as<NodeToSubstitute, ComponentVariableNode>
+                                         || std::same_as<NodeToSubstitute, PortFieldNode>;
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp:42]: You might use a broader concept here
```cpp
template<class T>
concept NodeConstructibleFromTwoStrings = std::constructible_from<T, std::string, std::string>
                                          && std::derived_from<T, Node>;
```

File: src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h [hunk 3]
```diff
@@ -31,28 +31,61 @@ namespace Antares::Solver::Visitors
...
+    {
+        return "SubstitutionVisitor";
+    }
 
 private:
     // Only override visit method for ComponentVariableNode, clone the rest
-    Nodes::Node* visit(const Nodes::ComponentVariableNode* node) override;
+    Nodes::Node* visit(const NodeToSubstitute* node) override
+    {
+        {
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h:77]: This scope level is useless, remove a pair of { }

File: src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp [hunk 4]
```diff
@@ -36,29 +37,37 @@ using namespace Antares::Solver::Visitors;
...
     };
-    return {add("component1", "variable1"), add("component2", "variable1")};
+    return {add("one", "two"), add("three", "four")};
 }
 
-BOOST_FIXTURE_TEST_CASE(SubstitutionVisitor_substitute_one_node, Registry<Node>)
+// Create a tuple of types
+typedef std::tuple<ComponentVariableNode, PortFieldNode> TestTypes;
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(SubstitutionVisitor_substitute_one_node, T, TestTypes)
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_SubstitutionVisitor.cpp:61]: nice !

File: src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h [hunk 5]
```diff
@@ -31,28 +31,61 @@ namespace Antares::Solver::Visitors
...
+template<class NodeToSubstitute>
 class SubstitutionVisitor: public CloneVisitor
 {
 public:
-    SubstitutionVisitor(Registry<Nodes::Node>& registry, SubstitutionContext& ctx);
+    SubstitutionVisitor(Registry<Nodes::Node>& registry,
+                        SubstitutionContext<NodeToSubstitute>& ctx):
+        CloneVisitor(registry),
+        ctx_(ctx),
+        registry_(registry)
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/visitors/SubstitutionVisitor.h:56]: The registry is only needed from `CloneVisitor`, `SubstitutionVisitor` doesn't use it directly. So you can remove `SubstitutionVisitor::registry_` and this line.

**Pull request closed without merging**